Add more info in the console runner + fix bug in trivia plugin

dvirZion opened the pull request.

dvirZion committed:
Improving the user interface, by adding a variable that tells the user if he is in the main menu or in a session of some plugin.

dvirZion committed:
Adding opening lines, as well as an option to return to the main menu, by typing the word main or home

YacovGold commented:
Please change the title of the PR to be more explanatory

YacovGold commented on Runners/ConsoleApp/Program.cs:
```diff
@@ -23,11 +24,14 @@ static void Main()
                 var msg = Console.ReadLine();
-                pluginExecutor.Run(msg, " ");
+                level=pluginExecutor.Run(msg, " ");
+               
```
Please format the code
https://docs.microsoft.com/en-us/answers/questions/550115/automatic-format-code-in-visual-studio.html

dvirZion replied:
ok

YacovGold replied:
If you fixed this, please click on the "Resolve conversation"

YacovGold commented on Runners/ConsoleApp/Program.cs:
```diff
@@ -23,11 +24,14 @@ static void Main()
             var pluginExecutor = new PluginExecutor(messageSender, new MemoryDal(), new PluginsMenu(), new PluginsManager());
-
+            Console.WriteLine("Hi! I am Your amazing bot and I will be happy to be of service to you on several topics.");
+            Console.WriteLine("Please type the number of the plugin you would like to help, or type help to get the list of options.\r\nAt any point you can type MAIN or HOME to return to the main menu.");
```
The msg here is long with \r\n
Instead, just use another WriteLine

dvirZion replied:
ok. It seems to me that the best thing would be to chain the messages with a small delay between them to give the feeling of a real chat. I just need more to learn how to do it in C#

YacovGold commented on Runners/ConsoleApp/Program.cs:
```diff
@@ -23,11 +24,14 @@ static void Main()
                 var msg = Console.ReadLine();
-                pluginExecutor.Run(msg, " ");
+                level=pluginExecutor.Run(msg, " ");
+               
```
Extra enter here

dvirZion replied:
I did not understand. Why call the function twice?

YacovGold replied:
I mean so there is an extra enter
You should remove it

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -31,8 +32,13 @@ public PluginExecutor(IMessageSender messageSender, IDal dal, PluginsMenu plugin
             _pluginsMenu = pluginsMenu;
             _pluginsManager = pluginsManager;
         }
+        public string PluginId()
```
This function is not in use, you can remove it

dvirZion replied:
ok. It was one of the attempts along the way.. I just forgot to delete.

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -42,16 +48,26 @@ public void Run(string message, string user)
-
                 Execute(pluginType, extraData, user);
+                currentPluginId = _dal.LoadPluginData(user, SESSION_PLUGIN_ID);
+                var level =(currentPluginId!=null)? "\\" + currentPluginId:"";
```
Instead of formatting the text here, we may return the current plugin id and the host (console\WhatsApp etc.) will decide how to format this

dvirZion replied:
ok

YacovGold commented:
If the system will use a real db, then, if the user start a session with a plugin and then close the app and launch again, it will be in a session, but the user will not know that. (you can check this by connecting to a real db, look at ChatbotContext->Data->ChatbotContext

dvirZion committed:
1. Correcting things as I was asked.
2. Fixing a bug that was in the code I added.
3. Adding a delay to the opening output

dvirZion commented:
Regarding the comment of the DB, as soon as there is a real DB, we can make a query and based on it update the LEVEL.

YacovGold commented:
> Regarding the comment of the DB, as soon as there is a real DB, we can make a query and based on it update the LEVEL.

There is :)
So we need to support this...

dvirZion committed:
Upgrading the system so that it also supports the use of the DB when displaying the add-on, already at the beginning of the activation.

dvirZion commented:
Upgrading the system so that it also supports the use of the DB when displaying the add-on, already at the beginning of the activation.

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -31,6 +33,11 @@ public PluginExecutor(IMessageSender messageSender, IDal dal, PluginsMenu plugin
             _pluginsMenu = pluginsMenu;
             _pluginsManager = pluginsManager;
         }
+        public string getPluginId(string user)
```
1. The function name should be in 'pascal case'
2. Please give the function more meaningful name (Something as 'GetCurrentUserPluginId')

dvirZion replied:
OK, I changed as you requested

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -42,16 +49,24 @@ public void Run(string message, string user)
                 {
                     _messageSender.SendMessage(user, msgForUser);
-                    return;
+                    return ;
```
Here you added an extra space, get use to format the code...

dvirZion replied:
OK

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -42,16 +49,24 @@ public void Run(string message, string user)
                 var pluginType = ExtractPluginType(pluginNumber);
-
                 Execute(pluginType, extraData, user);
+                currentPluginId = _dal.LoadPluginData(user, SESSION_PLUGIN_ID);
```
Is this line needed?

dvirZion replied:
No... I forgot to delete

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -42,16 +49,24 @@ public void Run(string message, string user)
             else
             {
-                Execute(currentPluginId, message, user);
+                if (message.ToLower() == "main" || message.ToLower() == "home")
```
You auto exit a session on this 2 keywords
Why so?
Maybe there is a plugin so ask user to type this keywords to do something?
If you override the functionality, it may make the system rigid :(

dvirZion replied:
ok i removed it

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -42,16 +49,24 @@ public void Run(string message, string user)
+                {
+                    Execute(currentPluginId, message, user);
+                }
+                currentPluginId = _dal.LoadPluginData(user, SESSION_PLUGIN_ID);
```
Is this line needed?

dvirZion replied:
No... I forgot to delete

YacovGold commented on Infrastructure/PluginExecutor.cs:
```diff
@@ -42,16 +49,24 @@ public void Run(string message, string user)
                     || CheckIfIlegalPluginPressed(message, out int pluginNumber, out msgForUser))
```
You created a function so get the current plugin, instead of loading it explicitly like this
`_dal.LoadPluginData(user, SESSION_PLUGIN_ID)`
You can call the function you created

This is important because in this way, you only need to change logic in one place if need to...

dvirZion replied:
OK

YacovGold commented on Runners/ConsoleApp/Program.cs:
```diff
@@ -23,14 +26,30 @@ static void Main()
             }
         }
 
+        private void OutputTextWithDelay(string txt)
```
This function is not in use, please rewove

dvirZion replied:
OK, I forgot to delete

YacovGold commented on Runners/ConsoleApp/Program.cs:
```diff
@@ -23,14 +26,30 @@ static void Main()
             {
+                var PluginId= pluginExecutor.getPluginId(" ");
+                PluginId = (PluginId!= null && PluginId!="") ? "\\" + PluginId : "";
+                Console.Write("main{0}> ", PluginId);
```
Please format the code...

dvirZion replied:
OK

dvirZion committed:
Correcting things as I was asked

dvirZion commented:
I made the corrections

dvirZion committed:
Fixing several bugs that were in the TRIVIA plugin

dvirZion commented:
Fixing several bugs that were in the TRIVIA plugin

dvirZion commented:
Fixing several bugs that were in the TRIVIA plugin

dvirZion commented:
Fixing several bugs that were in the TRIVIA plugin

YacovGold commented on Plugins/Trivia/TriviaPlugin.cs:
```diff
@@ -6,7 +6,7 @@
 
 namespace Trivia
 {
-    record PersistentDataStructure(double score, double numQest, int trueResult);
```
Fixing the bug is not related to this PR
You know how to work with branches?

dvirZion replied:
no

YacovGold replied:
I recommend learning this...
search "git branches"...

dvirZion replied:
OK. What about the current one's approval?

YacovGold replied:
For now, I will approve
